### ISSUE TITLE
`(require 'crafted-osx-config)` does nothing

### ISSUE THREAD

**Issue by jf** [SystemCrafters/crafted-emacs#415]
### Operating System

Mac

### Emacs Version

29.1

### Emacs Configuration Details

- [X] Native Compilation
- [ ] pGTK
- [ ] alternative package manager (use-package, straight.el, ...)

### Anything else that may be related to the issue you're having?

maybe this matters... and maybe it doesn't (I'm guessing it shouldnt really), but I'm using emacs-mac-app 10.0 (https://ports.macports.org/port/emacs-mac-app/), installed via macports

### What happened?

I've got crafted-emacs set up as per https://github.com/SystemCrafters/crafted-emacs/blob/master/docs/getting-started-guide.org, and seeing as how there is a `crafted-osx-config.el`, I go ahead to do `(require 'crafted-osx-config)`.

This seems to achieve nothing, though. Reading the code, I expect `⌘-t` to give me a new tab... instead it transposes words. I dont see a `crafted-osx-packages.el`... so I'm not expecting to have to install any packages to support the c onfig.

Minimal demo `init.el`:
```
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (and custom-file
	   (file-exists-p custom-file))
  (load custom-file nil))

(load "~/.config/crafted-emacs/modules/crafted-init-config")

(require 'crafted-osx-config)
```

### What should have happened?

see "What happened?"

**Comment by sthesing** [comment]
@jf Thanks for opening this. `crafted-osx-config` is one of the things we can't test very well ourselves.

The way we check whether a user is on MacOS is by checking for the feature `ns` (which stands for NextStep, I believe). An explanation why the module doesn't do anything might be that `(featurep 'ns)` evaluates to `nil` in your setup.
Can you check that for us? Maybe we just need to change the conditions in the code.

In the meantime, as a workaround, you can try to copy the contents of the module to your config and remove the `(when (featurep 'ns)` and the corresponding `)` after the wrapped code and see whether it works as intended.

**Comment by jf** [comment]
ok:

1. it looks like your macos detection is one fault here: evaluating `(featurep 'ns)` in a scratch buffer in Emacs gives me `nil`.
2. who wrote or put out this code then? this code really should have been tested.
3. no ELisp expert here, but a search shows that I can check `system-type`, and at least on my system, it works: `(eq system-type 'darwin)`

**Comment by sthesing** [comment]
Thanks for checking. Yeah, system type is an option. Problem is that different builds of Emacs for MacOS might not use the same feature-set as the ns-build. Can you check whether the following variables/functions exist on your system? E.g. by using "C-h o"?

- `mac-right-option-modifier`
- `mac-command-modifier`
- `ns-function-modifier`
- `ns-do-applescript`

**Comment by jf** [comment]
Doing that already.

`mac-right-option-modifier` and `mac-command-modifier` exist;
*NOT* so `ns-function-modifier` and `ns-do-applescript`!

**Comment by sthesing** [comment]
Thanks. Suspected as much. 
(With ns-build, I meant Emacs builds for MacOS where `(featurep 'ns)` evaluates to non-nil.)

For you (and for now), this means that you can set  `mac-command-modifier` to `super` and use the keybindings defined in the module, if you like them:

```elisp
(customize-set-variable mac-command-modifier 'super)

(keymap-global-set "s-W" #'delete-frame) ; ⌘-W = Close window
(keymap-global-set "s-}" #'tab-bar-switch-to-next-tab) ; ⌘-} = Next tab
(keymap-global-set "s-{" #'tab-bar-switch-to-prev-tab) ; ⌘-{ = Previous tab
(keymap-global-set "s-t" #'tab-bar-new-tab) ;⌘-t = New tab
(keymap-global-set "s-w" #'tab-bar-close-tab) ; ⌘-w = Close tab

(unless (version< emacs-version "28")
  (keymap-global-set "s-Z" 'undo-redo)) ; ⌘-Z = Redo
```

The rest of the module isn't of much use to your build of Emacs, I'm afraid.

For the future of the module, I feel we either need to find a way to accommodate the different flavours of Emacs for MacOS, or we document its limitations.

**Comment by jf** [comment]
thank you for the help. After reviewing and thinking about things, I think i'll stay away from this module. I've got the mac command key as meta already (emacs-mac-app does that, I believe)... and I like it that way.

**Comment by jeffbowman** [comment]
> who wrote or put out this code then? this code really should have been tested.

Just to point out, this code was provided by a community member who uses a Mac and Emacs, most likely either built by them or possibly from a `brew` installation.  It isn't the only way to get Emacs and different distributors' builds may affect how some things work.

All that said, I appreciate your engagement and discussion here, and perhaps you could help improve the code by submitting a patch (via Pull Request) ?

**Comment by jf** [comment]
re submitting a patch, I was actually about to when I discovered `system-type`; but paused after noticing the description for the `darwin` value (`compiled for Darwin (GNU-Darwin, macOS, ...).`). I mean... what is GNU-Darwin? do I even know what I'm doing?

I just looked into what GNU-Darwin is now, and I am thinking right now that `system-type` is probably the right fix; but the discussion and point about MacOS having different builds did make me think twice. My last thought on this is that it probably makes sense that whatever build anybody uses for the mac, it _should_ use the same value of `darwin` for `system-type`. Now if the *build maintainer* decides to mess with that, and change that... well that should be  on them.

Back to `crafted-osx-config.el` specifically: I'm looking back at the history of the file (starting from its genesis, I believe, as `rational-osx.el`).. and I'm not exactly getting why things were changed as they were along the way. For example https://github.com/SystemCrafters/crafted-emacs/commit/4095e3e442d0380ebf0c3aec8c39d89a82d8be50: why go from mapping "mac-command-modifier 'hyper", to "mac-command-modifier 'super"... just like that? and what makes those dropped keybindings redundant? does the ns-build provide "⌘-c = Copy" already? it's not altogether clear.

In summary:
- I can provide a PR to use `system-type` instead for os detection. I think this is the right fix
- I can*not* say that I am an expert at Emacs now, nor on the different options and builds available for the mac. But I am getting a better sense of things now, and I think I know what to do to at least have things work for somebody who is on the mac but does *not* have the ns-build, and who still wants to try out `crafted-osx-config`

**Comment by jf** [comment]
alright I have #416 now. Please have a look. I must say that it is disappointing what happened with the `customize-set-variable` calls.

One question though: would using `setq` be preferable? what is a `defcustom` value? custom variables that you define (and therefore not built-in variables like `mac-command-modifier`?)?

**Comment by jeffbowman** [comment]
>  I must say that it is disappointing what happened with the customize-set-variable calls.

Not sure exactly what you mean by this, so I'm not sure how to address it.

> One question though: would using setq be preferable? what is a defcustom value? 

`defcustom` is the standard way to create a customization property for Emacs.  `customize-set-variable` is the proper way to set the value of such a property because there may be additional code attached to the property which will only be run when this form is used to set the value.  `setq` is a macro that usually expands to `(set (quote var) value)` or something reasonably close to that.  It does not cause any additional code to run which might be attached to the `defcustom` property.   This is mentioned in the README.  Does this help answer the question?

See also: https://www.gnu.org/software/emacs/manual/html_node/emacs/Easy-Customization.html

**Comment by jf** [comment]
> > I must say that it is disappointing what happened with the customize-set-variable calls.
> 
> Not sure exactly what you mean by this, so I'm not sure how to address it.
> 

I am referring to https://github.com/SystemCrafters/crafted-emacs/blob/449bde3f35cd240225882d2e2e8a692059311cf5/modules/crafted-osx-config.el#L24-L26, and how initially, pre- #416, it looks like, unless Emacs changed its behaviour for `customize-set-variable`, these calls never worked (because `mac-command-modifier`, etc. needs to be quoted). Which means unless you had your variables set *already* to the right values, this module would not work. ⌘-W would not "Close window", etc.

We do need more eyes on the code, so I appreciate your move to call on others to have a look as well for #416.

> > One question though: would using setq be preferable? what is a defcustom value?
> 
> `defcustom` is the standard way to create a customization property for Emacs. `customize-set-variable` is the proper way to set the value of such a property because there may be additional code attached to the property which will only be run when this form is used to set the value. `setq` is a macro that usually expands to `(set (quote var) value)` or something reasonably close to that. It does not cause any additional code to run which might be attached to the `defcustom` property. This is mentioned in the README. Does this help answer the question?
> 
> See also: https://www.gnu.org/software/emacs/manual/html_node/emacs/Easy-Customization.html

thank you, that helps. My question also relates to (I assume) "built-in" variables like `mac-command-modifier`, vs variables that we define ourselves through `defcustom` (like `crafted-startup-recentf-count`, for instance). Is it also preferable to use `customize-set-variable` for these built-in variables? does the rationale for using `customize-set-variable` also apply here?

**Comment by jeffbowman** [comment]
> I just did a bit more reading and I think i get it now. Could you help me confirm my understanding? If I see in the help that "You can customize this variable" I should be defaulting to customize-set-variable. Is this correct?

Yes, that is correct.

**Comment by sthesing** [comment]
Closing this, as it's addressed by #416 . Feel free to reopen if I'm wrong.
